mmdet3d/ops/paconv/paconv.py

```diff
@@ -0,0 +1,350 @@
+                 bias='auto'):
+        super(ScoreNet, self).__init__()
+
+        assert score_norm in ['softmax', 'sigmoid', 'identity'], \
+            f'unsupported score_norm function {score_norm}'
+
+        self.score_norm = score_norm
+        self.temp_factor = temp_factor
+
+        self.mlps = nn.Sequential()
+        for i in range(len(mlp_channels) - 2):
+            self.mlps.add_module(
+                f'layer{i}',
+                ConvModule(
+                    mlp_channels[i],
+                    mlp_channels[i + 1],
+                    kernel_size=(1, 1),
+                    stride=(1, 1),
+                    conv_cfg=dict(type='Conv2d'),
+                    norm_cfg=norm_cfg,
+                    bias=bias))
+
+        # for the last mlp that outputs scores, no relu and possibly no bn
+        i = len(mlp_channels) - 2
+        self.mlps.add_module(
+            f'layer{i}',
+            ConvModule(
+                mlp_channels[i],
+                mlp_channels[i + 1],
+                kernel_size=(1, 1),
+                stride=(1, 1),
+                conv_cfg=dict(type='Conv2d'),
+                norm_cfg=norm_cfg if last_bn else None,
+                act_cfg=None,
+                bias=bias))
+
+    def forward(self, xyz_features):
+        """Forward.
+
+        Args:
+            xyz_features (torch.Tensor): (B, C, N, K), features constructed
+                from xyz coordinates of point pairs. May contain relative
+                positions, Euclidian distance, etc.
+
+        Returns:
+            torch.Tensor: (B, N, K, M), predicted scores for `M` kernels.
+        """
+        scores = self.mlps(xyz_features)  # (B, M, N, K)
+
+        # perform score normalization
+        if self.score_norm == 'softmax':
+            scores = F.softmax(scores / self.temp_factor, dim=1)
+        elif self.score_norm == 'sigmoid':
+            scores = torch.sigmoid(scores / self.temp_factor)
+        else:
+            scores = scores
+
+        scores = scores.permute(0, 2, 3, 1)  # (B, N, K, M)
+
+        return scores
+
+
+class PAConv(nn.Module):
+    """Non-CUDA version of PAConv.
+
+    PAConv stores a trainable weight bank containing several weight kernels.
+    Given input points and features, it computes coefficient scores to assemble
+    those kernels to form conv kernels, and then runs convolution on the input.
+
+    Args:
+        in_channels (int): Input channels of point features.
+        out_channels (int): Output channels of point features.
+        num_kernels (int): Number of weight kernels in the weight bank.
+        norm_cfg (dict, optional): Type of normalization method.
+            Defaults to dict(type='BN2d', momentum=0.1).
+        act_cfg (dict, optional): Type of activation method.
+            Defaults to dict(type='ReLU', inplace=True).
+        scorenet_input (str, optional): Type of input to ScoreNet.
+            Can be 'identity', 'neighbor' or 'ed7'. Defaults to 'ed7'.
```

Comment:
ed7 is quite strange, and 'neighbor' actually refers to 'identity'+'neighbor'?

Reply:
Changed.